src/rpc/AppRpcProvider.ts

```diff
@@ -0,0 +1,156 @@
+
+function now() {
+  return new Date().getTime()
+}
+
+function checkNetworks(networks: Array<Network>): Network | null {
+  let result: Network | null = null
+
+  for (let i = 0; i < networks.length; i++) {
+    const network = networks[i]
+
+    // Null! We do not know our network; bail.
+    if (network == null) {
+      return null
+    }
+
+    if (result) {
+      // Make sure the network matches the previous networks
+      if (
+        !(
+          result.name === network.name &&
+          result.chainId === network.chainId &&
+          (result.ensAddress === network.ensAddress || (result.ensAddress == null && network.ensAddress == null))
+        )
+      ) {
+        return null
+      }
+    } else {
+      result = network
+    }
+  }
+
+  return result
+}
+
+interface ProviderPerformance {
+  latency: number
+  failureRate: number
+  lastEvaluated: number
+}
+
+interface FallbackProviderEvaluation {
+  provider: JsonRpcProvider
+  performance: ProviderPerformance
+}
+
+/**
+ * AppRpcProvider is an extension of the JsonRpcProvider class from the ethers.js library.
+ * This provider balances requests among multiple JSON-RPC endpoints.
+ */
+export default class AppRpcProvider extends JsonRpcProvider {
+  readonly providerEvaluations: ReadonlyArray<FallbackProviderEvaluation>
+  readonly evaluationInterval: number
+
+  constructor(providers: JsonRpcProvider[], evaluationInterval = 30000) {
+    if (providers.length === 0) throw new Error('providers array empty')
+    providers.forEach((provider, i) => {
+      if (!Provider.isProvider(provider)) throw new Error(`invalid provider ${i}`)
+    })
+    const agreedUponNetwork = checkNetworks(providers.map((p) => p.network))
+    if (!agreedUponNetwork) throw new Error('networks mismatch')
+
+    super(providers[0].connection)
+    this.providerEvaluations = providers.map((provider) => {
+      return Object.freeze({
+        provider,
+        performance: {
+          latency: Number.MAX_SAFE_INTEGER,
+          failureRate: 0,
+          lastEvaluated: 0,
+        },
+      })
+    })
+
+    this.evaluationInterval = evaluationInterval
+  }
+
+  /**
+   * Perform a JSON-RPC request.
+   *
+   * method - The JSON-RPC method name.
+   * params - The parameters for the JSON-RPC method.
+   * Returns a Promise that resolves with the result of the JSON-RPC method.
+   * Throws an error if all providers fail to perform the operation.
+   */
+  async perform(method: string, params: { [name: string]: any }): Promise<any> {
+    // Periodically evaluate all providers
+    const currentTime = now()
+    this.providerEvaluations.forEach((providerEval) => {
+      if (currentTime - providerEval.performance.lastEvaluated >= this.evaluationInterval) {
+        this.evaluateProvider(providerEval)
+      }
+    })
+
+    // Sort providers by performance score (latency and failure rate)
+    const sortedEvaluations = AppRpcProvider.sortProviders(this.providerEvaluations.slice())
```

Comment:
Sort as part of evaluateProvider every time an evaluation comes in, so that you're not re-sorting every time you call `perform`.